### PR TITLE
Fix bugs in the selenium tests from "factory" package.

### PR DESCRIPTION
**Pull request by Ohrimenko1988**
### What does this PR do?
Fix bugs in the selenium tests from "factory" package. Fixed problem was  very closely related with another bugs like problem with open item by path in project explorer, bug with login page and other.  
So this PR fix several issue.

*Changes were tested on CI:*
*last 3 runs:*
https://ci.codenvycorp.com/view/qa/job/experimental-che-multiuser-integration-tests-che6/13/Selenium_tests_report/

https://ci.codenvycorp.com/view/qa/job/experimental-che-multiuser-integration-tests-che6/11/Selenium_tests_report/

https://ci.codenvycorp.com/view/qa/job/experimental-che-multiuser-integration-tests-che6/10/Selenium_tests_report/

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/7340
https://github.com/eclipse/che/issues/7227
https://github.com/eclipse/che/issues/7387

### CODE REVIEW

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/ProjectExplorer.java [hunk 1]
```diff
@@ -140,6 +144,16 @@ public ProjectExplorer(
   @FindBy(xpath = Locators.PROJECT_EXPLORER_TAB_IN_THE_LEFT_PANEL)
   WebElement projectExplorerTabInTheLeftPanel;
 
+  /**
+   * @param path path to item in Project Explorer
+   * @return item by path
+   */
+  private WebElement getProjectExplorerItem(String path) {
+    String locator = "//div[@path='/%s']/div";
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/ProjectExplorer.java]: This variable seems is redundant here. Let's get rid of it to simplify the code.

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/ProjectExplorer.java [hunk 2]
```diff
@@ -370,8 +397,20 @@ public void selectItem(String path) {
...
     String locator = "//div[@path='/" + path + "']/div";
     waitItem(path, timeoutForWaitingItem);
-    WebElement item = seleniumWebDriver.findElement(By.xpath(locator));
-    item.click();
+    try {
+      new WebDriverWait(seleniumWebDriver, EXPECTED_MESS_IN_CONSOLE_SEC)
+          .until(ExpectedConditions.visibilityOfElementLocated(By.xpath(locator)))
+          .click();
+    } catch (StaleElementReferenceException ex) {
+      LOG.error(ex.getLocalizedMessage(), ex);
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/ProjectExplorer.java]: IMHO it's better to log exception as warn() here so as we don't interrupt the execution of tests.

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/ProjectExplorer.java [hunk 3]
```diff
@@ -370,8 +397,20 @@ public void selectItem(String path) {
   public void selectItem(String path, int timeoutForWaitingItem) {
     String locator = "//div[@path='/" + path + "']/div";
     waitItem(path, timeoutForWaitingItem);
-    WebElement item = seleniumWebDriver.findElement(By.xpath(locator));
-    item.click();
+    try {
+      new WebDriverWait(seleniumWebDriver, EXPECTED_MESS_IN_CONSOLE_SEC)
+          .until(ExpectedConditions.visibilityOfElementLocated(By.xpath(locator)))
+          .click();
+    } catch (StaleElementReferenceException ex) {
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/ProjectExplorer.java]: Such cases should be commented with short explanation of workaround, as a rule, to explain why we can ignore an exception and can try to repeat operation again.

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/ProjectExplorer.java [hunk 4]
```diff
@@ -428,24 +469,27 @@ public void scrollAndSelectItem(String path) {
...
+      getProjectExplorerItem(path).click();
       waitItemIsSelected(path);
-      actionsFactory.createAction(seleniumWebDriver).doubleClick(item).perform();
+      action.moveToElement(getProjectExplorerItem(path)).perform();
+      action.doubleClick().perform();
     }
     // sometimes an element in the project explorer may be is not attached to the DOM. We should
     // refresh all items.
     catch (StaleElementReferenceException ex) {
+      LOG.error(ex.getLocalizedMessage(), ex);
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/ProjectExplorer.java]: IMHO it's better to log exception as warn() here so as we don't interrupt the execution of tests.

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/ProjectExplorer.java (hunk 1 above)
Comment (Ohrimenko1988) [reply selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/ProjectExplorer.java]: We use this method 7 times, so on my opinion, this is better apply this method.

File: selenium/che-selenium-test/src/test/java/org/eclipse/che/selenium/factory/CheckFactoryWithSincePolicyTest.java [hunk 5]
```diff
@@ -66,10 +67,16 @@ public void checkFactoryAcceptingWithSincePolicy() throws Exception {
...
+    if (System.currentTimeMillis() > initTime + ADDITIONAL_TIME) {
+      Assert.fail(
+          "Factory started longer then additional time and next test steps does not make sense");
+    }
+
     warningDialog.waitWaitWarnDialogWindowWithSpecifiedTextMess(EXPIRE_MESSAGE);
 
     // wait until factory becomes avaialble
-    while (System.currentTimeMillis() <= INIT_TIME + ADDITIONAL_TIME) {
+    while (System.currentTimeMillis() <= initTime + ADDITIONAL_TIME + 10000) {
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/test/java/org/eclipse/che/selenium/factory/CheckFactoryWithSincePolicyTest.java]: **10000** needs to be named.

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/ProjectExplorer.java [hunk 6]
```diff
@@ -353,9 +367,22 @@ public void waitItemIsNotPresentVisibleArea(String item) {
...
     String locator = "//div[@path='/" + path + "']/div";
+
     waitItem(path);
-    WebElement item = seleniumWebDriver.findElement(By.xpath(locator));
-    item.click();
+    try {
+      new WebDriverWait(seleniumWebDriver, EXPECTED_MESS_IN_CONSOLE_SEC)
+          .until(ExpectedConditions.visibilityOfElementLocated(By.xpath(locator)))
+          .click();
+    } catch (StaleElementReferenceException ex) {
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/ProjectExplorer.java]: Such cases should be commented with short explanation of workaround, as a rule, to explain why we can ignore an exception and can try to repeat operation again.

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/ProjectExplorer.java [hunk 7]
```diff
@@ -353,9 +367,22 @@ public void waitItemIsNotPresentVisibleArea(String item) {
...
+      LOG.error(ex.getLocalizedMessage(), ex);
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/ProjectExplorer.java]: IMHO it's better to log exception as warn() here so as we don't interrupt the execution of tests.

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/ProjectExplorer.java [hunk 8]
```diff
@@ -373,8 +373,11 @@ public void selectItem(String path) {
       new WebDriverWait(seleniumWebDriver, EXPECTED_MESS_IN_CONSOLE_SEC)
           .until(ExpectedConditions.visibilityOfElementLocated(By.xpath(locator)))
           .click();
-    } catch (StaleElementReferenceException ex) {
-      LOG.error(ex.getLocalizedMessage(), ex);
+    }
+    // sometimes an element in the project explorer may be is not attached to the DOM. We should
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/ProjectExplorer.java]: correct variant is `may not be attached`

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/ProjectExplorer.java [hunk 9]
```diff
@@ -401,8 +404,11 @@ public void selectItem(String path, int timeoutForWaitingItem) {
...
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/ProjectExplorer.java]: correct variant is `may not be attached`.
Could you, please, fix it in the whole class?

File: selenium/che-selenium-test/src/test/java/org/eclipse/che/selenium/factory/CheckFactoryWithSincePolicyTest.java [hunk 10]
```diff
@@ -35,7 +35,8 @@
   private static final String EXPIRE_MESSAGE =
       "Unable to load Factory: This Factory is not yet valid due to time restrictions applied"
           + " by its owner. Please, contact owner for more information.";
-  private static final int ADDITIONAL_TIME = 120000;
+  private static final int FACTORY_IN_ACTIVITY_TIME = 120000;
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/test/java/org/eclipse/che/selenium/factory/CheckFactoryWithSincePolicyTest.java]: `FACTORY_INACTIVITY_TIMEOUT` is more suitable here.

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/ProjectExplorer.java [hunk 11]
```diff
@@ -232,7 +232,7 @@ public void waitItem(String path) {
    * @param timeout user timeout
    */
   public void waitItem(String path, int timeout) {
-    String locator = "//div[@path='/" + path + "']";
+    String locator = "//div[@path='/" + path + "']/div";
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/ProjectExplorer.java:235]: Could you, please, inline all such variables in class, to simplify the code?

**Comment by dmytro-ndp** [comment]
FYI: this issue seems partly fixes #6524.
Hope, the changes will be merged into che5.x branch as well.

**Comment by dmytro-ndp** [comment]
Also, to simplify review, it's better not to unite fixes for different issues into the one pull request in advance.

**Review by dmytro-ndp — APPROVED** [review]

**Review by tolusha — APPROVED** [review]

**Pull request merged**